lighthouse-core/gather/gatherers/iframe-elements.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('./gatherer.js');
+const pageFunctions = require('../../lib/page-functions.js');
+
+/* eslint-env browser, node */
+
+/**
+ * @return {LH.Artifacts['IFrameElements']}
+ */
+/* istanbul ignore next */
+function collectIFrameElements() {
+  // @ts-ignore - put into scope via stringification
+  const iFrameElements = getElementsInDocument('iframe'); // eslint-disable-line no-undef
+  return iFrameElements.map(/** @param {HTMLIFrameElement} node */ (node) => {
+    const clientRect = node.getBoundingClientRect();
+    const {top, bottom, left, right, width, height} = clientRect;
+    return {
+      id: node.id,
+      src: node.src,
+      clientRect: {top, bottom, left, right, width, height},
+      // @ts-ignore - put into scope via stringification
+      isPositionFixed: isPositionFixed(node), // eslint-disable-line no-undef
+    };
+  });
+}
+
+class IFrameElements extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['IFrameElements']>}
+   * @override
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    const {frameTree} = await driver.sendCommand('Page.getFrameTree');
+    let toVisit = [frameTree];
+    /** @type {Map<string | undefined, LH.Crdp.Page.Frame | undefined>} */
+    const framesByDomId = new Map();
+
+    while (toVisit.length) {
+      const tempFrameTree = toVisit.shift();
+      if (tempFrameTree) {
+        for (const childFrameTree of tempFrameTree.childFrames || []) {
+          if (childFrameTree.childFrames) {
+            toVisit = toVisit.concat(childFrameTree.childFrames);
+          }
+          if (framesByDomId.has(childFrameTree.frame.name)) {
```

Comment:
nit: feels a bit confusing to do all this grandchildren work instead of just letting the while loop take care of the 2nd level.

WDYT about

```js
while (toVisit.length) {
  const frameTree = toVisit.shift()
  // Shouldn't be necessary but make tsc happy
  if (!frameTree) continue;
  // Set the dom ID
  if (framesByDomId.has(frameTree.frame.name)) {
    framesByDomId.set(frameTree.frame.name, undefined);
  } else {
    framesByDomId.set(frameTree.frame.name, frameTree.frame);
  }

  // Add all the children to our queue
  if (frameTree.childFrames) {
    toVisit = toVisit.concat(frameTree.childFrames);
  }
}